### DISCUSSION TITLE
TextLog Max Lines Behaviour

### DISCUSSION THREAD

**Discussion by mattloose** [Textualize/textual#2561] [General]
Hi,

I am using the TextLog widget and I want to use it to provide a recent subset of information that is being written to a log file.

Becuase it will be used in code that is logging information up to 5 times per second, I want to limit the length of the log using the max lines behaviour.

However, I was expecting max lines to work in a fifo manner - i.e you could keep updating beying the line limit and it would pop out the oldest entry to replace it. It doesn't appear to work that way - rather the log seems to be limited at that length and no further info can be added. Am I correct and is there a suggested way of having a fifo type log?

Thanks for a great tool!

**Comment by davep** [comment]
I just tested `TextLog` as of 0.24.1 with this code:

```python
from textual.app        import App, ComposeResult
from textual.containers import Grid
from textual.widgets    import TextLog
from textual.reactive   import var

class TextLogMaxLinesApp( App[ None ] ):

    CSS = """
    Grid {
        grid-size: 3;
    }

    TextLog {
        border: solid yellow;
    }
    """

    line: var[ int ] = var( 0 )

    def compose( self ) -> ComposeResult:
        with Grid():
            for _ in range( 9 ):
                yield TextLog( max_lines=10 )

    def watch_line( self ) -> None:
        for log in self.query( TextLog ):
            log.write( str( self.line ) )

    def next_line( self ) -> None:
        self.line += 1

    def on_mount( self ) -> None:
        self.set_interval( 1 / 5, self.next_line )

if __name__ == "__main__":
    TextLogMaxLinesApp().run()
```

and I'm seeing the behaviour you're expecting to see:

https://github.com/Textualize/textual/assets/28237/1d2ac3b5-bf8b-416b-ad60-b0340d7dd51a

**Reply by mattloose** [reply]
Ok - thanks for the quick response - let me go back and double check my code!

**Reply by davep** [reply]
Did you get to the bottom of this?

**Reply by mattloose** [reply]
Yes - I did.

There were two issues - I had a silent fail in the code sending information to the log that would cause it to stop updating at unpredictable time points. Then the underlying process that generated the logging data itslef will intermittently pause every few minutes. The two combined made it appear that the log woud fill up and I assumed the log was failing. Adding a timestamp to every message (and your clear demonstration of the correct behaviour!) helped me track those issues. Thank you!

**Reply by davep** [reply]
Cool, glad you got it sorted. :-)
